### DISCUSSION TITLE
Feature proposal / RFC: sidenotes

### DISCUSSION THREAD

**Discussion by notpushkin** [squidfunk/mkdocs-material#4688] [Ideas]
<img width="1552" alt="image" src="https://user-images.githubusercontent.com/1298948/204470720-387def21-9c1e-45e1-b119-ec0140da4b5e.png">

See it live: https://material-tufte.ale.sh/
Source code: https://material-tufte.ale.sh/stylesheets/tufte-sidebars.css

**Comment by squidfunk** [comment]
Excellent idea!

**Comment by notpushkin** [comment]
An example in the wild (with the Blog addon!):

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/1298948/206345328-3b512c72-ef85-4838-a8f5-4b307cff48eb.png">

Still trying to get it working with footnotes :')

**Comment by notpushkin** [comment]
Another example:

<img width="1550" alt="image" src="https://user-images.githubusercontent.com/1298948/208053443-59f025dc-d35e-4a4d-a961-a512c5bc8c16.png">

This could work pretty well as an alternative view for code annotations if implemented properly (right now this is a pile of CSS hacks making a few code blocks look like a continuous one).

UPD: see it live: https://lunni.aedge.dev/deploy/

**Reply by squidfunk** [reply]
Definitely interesting. How do you imagine this working on mobile? Does this mandate the use of `toc.integrate`?

**Reply by notpushkin** [reply]
@squidfunk For code blocks in particular, I'd say it can just fall back to click-to-show annotations on mobile? While I generally prefer interleaving notes with main content, breaking in the middle of a code listing might look a bit strange:

<img width="375" alt="image" src="https://user-images.githubusercontent.com/1298948/208072885-de8150e4-a88d-47ea-9786-a8af0ba611a7.png">

**Reply by notpushkin** [reply]
> Does this mandate the use of `toc.integrate`?

I'd say yeah, either that or disabling TOC altogether. Unless you have any ideas how to make them work together? :-)

**Reply by squidfunk** [reply]
Disabling table of contents is probably not an option for many users and `toc.integrate` has several short comings, i.e. it doesn't work well for blogs (yet?) and can't be combined with section index pages. From the top of my head, mobile support and table of contents are the main two blockers that need some deep thinking before tackling sidenotes. I like the idea of sidenotes and think there's value in providing it as a feature, but it's definitely not easy to implement properly.

**Comment by notpushkin** [comment]
Let's discuss mobile support here.

From the top of my head, I see three possible uses of sidenotes:

## Plain margin notes

Just some text on the margin. On mobile, those can just be inlined after the paragraph they're attached to:

<details>
<summary>Screenshot</summary>
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/1298948/208111250-d34c9437-91f3-4b58-954a-cefd8272496f.png">
</details>

Alternatively, Tufte CSS implements those as a ⊕ link/button inside the paragraph, but I see this as undesirable: it requires another tap on a rather small button to show content, rather than allowing to just scroll past through it. Same goes for tap-to-show footnotes.


## Footnotes

Those would too look nice inlined on mobile IMHO:

<details>
<summary>Screenshot</summary>
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/1298948/208113823-ae6a04fa-3b83-492d-beec-2c9bfeae8013.png">
</details>

## Code annotations

Compared to footnotes, which have numbers and thus can be safely moved a _bit_ away from the related text, code annotations are unnumbered and should occur near the line they are talking about so that it is easier for the reader to follow.

On mobile, we can just reuse the current code annotation logic with the tap-to-show button. Some alternative approaches:

- Just showing all the annotations below code block is suboptimal because code blocks that need annotations are probably long enough and user will have to scroll back and forth.
- Interleaving code with annotations ([can be seen here][q]) is hard to implement properly if we have a single `<pre><code>` block.

[q]: https://github.com/squidfunk/mkdocs-material/discussions/4688#discussioncomment-4419964

**Comment by notpushkin** [comment]
Let's discuss TOC in this thread.

I don't really see a way to integrate right-hand TOC with sidenotes _cleanly_.

- We could make TOC static, or fixed on the screen until it bumps into the first note. We'll need to make sure there's no note collapsing with it on the very start. (how?)

- Less complicated solution could be to allow TOC prefixing the content itself. Perhaps this could be a good option for blog posts? I'd make it opt-in though, as it only makes sense for longer posts.

Any other ideas?

**Reply by squidfunk** [reply]
1. This will not work with the current architecture (it would need to be sticky and sidenotes would need to live in the same container, which they don't. I'm not sure I can even imagine this working without making it overly complex. Code annotations are already quite complex.
2. is fundamentally different from what we have now. It's definitely possible to pull off, but I'm not sure we want to support a third ay to render the table of contents besides on the right and on the left. Thus I'd rather leave this option as an exercise to the reader 😊

Thus I'm also not really seeing it. Maybe you can fiddle around a little and see if you can pull something off that basically works. We could then discuss whether this might be a good fit integrating. At this state I'd view it as a pre-prototype.

**Reply by notpushkin** [reply]
I'll give it some more thought. Thanks for the feedback!

Perhaps there's a way to give this a tiny bit of visibility? I feel like this problem could use another pair of eyes ✨

**Reply by squidfunk** [reply]
I've pinned the discussion, maybe some other users have some ideas. In order to convert it to a feature request and consider it for implementation, we would need a solid proposal. For this to happen, some thought is necessary. To be clear: I'm not yet convinced that the mechanics generalize well enough. Happy to be convinced, though, as I like the idea.
